p2p/server.go

```diff
@@ -899,18 +895,23 @@ func (srv *Server) listenLoop() {
 				time.Sleep(time.Millisecond * 200)
 				continue
 			} else if err != nil {
-				srv.log.Trace("Read error", "err", err)
-				slots <- struct{}{}
+				// Log the error unless the server is shutting down.
+				select {
+				case <-srv.quit:
+				default:
+					srv.log.Error("Server listener failed to accept a connection", "err", err)
+				}
+				slots.Release(1)
 				return
 			}
 			break
 		}
 
 		remoteIP := netutil.AddrIP(fd.RemoteAddr())
 		if err := srv.checkInboundConn(fd, remoteIP); err != nil {
-			srv.log.Trace("Rejected inbound connnection", "addr", fd.RemoteAddr(), "err", err)
-			fd.Close()
-			slots <- struct{}{}
+			srv.log.Trace("Rejected inbound connection", "addr", fd.RemoteAddr(), "err", err)
+			_ = fd.Close()
+			slots.Release(1)
```

Comment:
better do release in defer, otherwise can get deadlock on panic

Reply:
I'm not sure how to do this without a bigger refactoring of the structure of this function.

defer only runs on return, and can't be cancelled once it is set.
Here the logic wants to release either before goroutine if checks fail, or after the setup goroutine ends.

One way to do would be like so (pseudocode):
```
acquire()
preSetupBarrier = WaitGroup(1)
go {
    defer release()
    conn = accept()
    if err = checks(conn); err != nil { return }
    preSetupBarrier.Done()
    setup(conn)
}
preSetupBarrier.Wait()
```

it is definitely neat and future-proof with a single Release
but log.Trace and fd.Close are primitives and are unlikely to panic.

Do you think it is worth this refactoring?

Also I wonder if we'd like to keep the ability to apply patches from geth upstream?

Reply:
maybe it could be better refactored with a bounded channel of accepted connections that need to be set up?